Add GPTNeoXCausalLM

shivance opened the pull request:
We don't have seperate q, k, v layers in GPT NeoX, how do we handle caching logic in Decoder and Attention?


In favor of #1052

shivance commented:
All tests pass, thanks for help @mattdangerw !

mattdangerw commented:
Probably we should land #1111 then this. It would also be good to verify the correctness of our output, by converting weights for a small checkpoint, e.g. the tiniest pythia, then running through this workflow. Have we tried that out?

shivance commented:
JAX and Torch tests are failing with seperate errors

mattdangerw commented:
@shivance yeah these look separate and worth looking into on their own.

For jax, someone one of our functions that should be stateless is not actually stateless, I'm not sure where a leak would have some in. If you can reproduce locally, you could try the `JAX_CHECK_TRACER_LEAKS` environment variable and use that.

The torch error looks a bit simpler. On this line `x1, x2 = ops.split(tensor, 2, axis=-1)`, the tensor passed in has shape `(2, 8, 2, 0)`, that last zero shouldn't be there. I also didn't not 0 shapes where supported, is this a torch thing. If you can figure how that axis is getting added and remove it, we are probably all set.

Also, a simple bug, but you need to reformat your code.

shivance commented:
An interesting revelation on behavior of split method across torch, tf and numpy

```
import torch
a = torch.ones(2, 8, 2, 0)
x = torch.split(a, 2, dim=-1)
print("splits = ", x)
```

output 
splits =  (tensor([], size=(2, 8, 2, 0)),)

```
import tensorflow as tf
b = tf.ones((2, 8, 2, 0))
y = tf.split(b, 2, axis=-1)
print("splits = ", y)
```

output 
splits =  [<tf.Tensor: shape=(2, 8, 2, 0), dtype=float32, numpy=array([], shape=(2, 8, 2, 0), dtype=float32)>, <tf.Tensor: shape=(2, 8, 2, 0), dtype=float32, numpy=array([], shape=(2, 8, 2, 0), dtype=float32)>]

```
import numpy as np

c = np.ones((2, 8, 2, 0))
z = np.split(c, 2, axis=-1)
print("splits = ", z)
```

output 
splits =  [array([], shape=(2, 8, 2, 0), dtype=float64), array([], shape=(2, 8, 2, 0), dtype=float64)]

I believe this different behavior of `split` causes error for PyTorch backend in multi-backend for this test case!

shivance commented:
@mattdangerw , regarding memory leak in JAX

here are last few stack traces

    return self.call(*args, **kwargs)
gpt_neo_x_attention.py:153: in call
    query_key_value = self._qkv_dense(hidden_states)
/home/shivance/anaconda3/envs/backend/lib/python3.10/site-packages/keras_core/src/utils/traceback_utils.py:114: in error_handler
    return fn(*args, **kwargs)
/home/shivance/anaconda3/envs/backend/lib/python3.10/site-packages/keras_core/src/layers/layer.py:703: in __call__
    outputs = super().__call__(*args, **kwargs)
/home/shivance/anaconda3/envs/backend/lib/python3.10/site-packages/keras_core/src/utils/traceback_utils.py:114: in error_handler
    return fn(*args, **kwargs)
/home/shivance/anaconda3/envs/backend/lib/python3.10/site-packages/keras_core/src/ops/operation.py:46: in __call__
    return self.call(*args, **kwargs)
/home/shivance/anaconda3/envs/backend/lib/python3.10/site-packages/keras_core/src/layers/core/einsum_dense.py:186: in call
    x = ops.einsum(self.equation, inputs, self.kernel)
/home/shivance/anaconda3/envs/backend/lib/python3.10/site-packages/keras_core/src/ops/numpy.py:1276: in einsum
    return backend.numpy.einsum(subscripts, *operands)
/home/shivance/anaconda3/envs/backend/lib/python3.10/site-packages/keras_core/src/backend/jax/numpy.py:26: in einsum
    return jnp.einsum(subscripts, *operands, **kwargs)
/home/shivance/anaconda3/envs/backend/lib/python3.10/site-packages/jax/_src/numpy/lax_numpy.py:3248: in einsum
    return _einsum_computation(operands, contractions, precision,  # type: ignore[operator]
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 


It points to the Einsumdense layer, and that's the place we are probably having memory leaks.

mattdangerw commented:
The jax error looks like a legitimate bug to me, I pushed a fix -> 
https://github.com/keras-team/keras-nlp/pull/1110/commits/09127f9a25e8db0ef874ceadb1284fd410524f39

Basically the rule with jax is to make sure all your variables are created in a layer `__init__` or `build`. Otherwise, you will break the stateless assumptions made by the library when attempting to compile a function.

Also note you can go ahead and ignore the accelerator errors for now.

shivance commented:
Awesome, Thanks for the help here @mattdangerw.
I'm trying to put pieces up for generate pipeline for `pythia-70m` checkpoint .

Here is the [colab](https://colab.research.google.com/drive/1aUfhmyeufGyKgWC4tvtLWcO6_5E0YBw5?usp=sharing) link. Could you please review if the flow is correct? 
Again, as of now, I am getting wrong output (i.e. output of generate is same as input LoL)

mattdangerw commented:
Nice work on the colab! The overall flow does look correct to me.

I would suggest first comparing the backbone output directly on some simple inputs to make sure that is 1-1, before looking at the whole generative sampling bit. That could help us verify that the weights are being converted correctly.

mattdangerw reviewed (commented):
Looks good overall! Some misc comments while looking over.

shivance commented on keras_nlp/models/gpt_neo_x/gpt_neo_x_causal_lm.py:
```diff
@@ -0,0 +1,232 @@
+from keras_nlp.utils.python_utils import classproperty
+
+
+class ReverseEmbedding(keras.layers.Layer):
```
They are different

shivance commented on keras_nlp/models/gpt_neo_x/gpt_neo_x_attention.py:
```diff
@@ -165,7 +190,8 @@ def call(
         )
 
-        query_rot, key_rot = self.rotary_embedding(query_rot, key_rot)
+        query_rot = self.rotary_embedding(query_rot)
```
Thanks !

shivance commented:
@mattdangerw I addressed all these comments. We could get this in and I'll continue to work on fixing generate as a followup PR.

mattdangerw commented:
/gcbrun

mattdangerw commented:
Pushed a rebased version, to get rid of the 40 file diff. You could update your local branch to this exactly by doing something like...

```
git checkout casual_lm
git fetch
git reset --hard origin/causual_lm
```

mattdangerw commented:
/gcbrun

mattdangerw commented:
/gcbrun

mattdangerw commented:
Thanks! Merging. Though IIUC we probably need some more bug fixing before our generative pass is totally right here.

shivance commented:
> Thanks! Merging. Though IIUC we probably need some more bug fixing before our generative pass is totally right here.

Got it ! Working on it.
